compiler/resolution/cullOverReferences.cpp

```diff
@@ -582,746 +582,848 @@ bool isRefOrTupleWithRef(Symbol* index, int tupleElement)
 //
-// The compiler must also determine whether the "by value" implementation
-// will have inserted an autoCopy.   If so then the compiler attaches
-// the AUTO_COPY/AUTO_DESTROY flags as necessary to enable the
-// callDestructors pass to operate correctly.
-//
-// This pass also concerns itself with lowering ArgSymbols with
-// INTENT_REF_MAYBE_CONST to INTENT_REF or INTENT_CONST_REF, depending
-// on whether the formal was "set" in the body of the function.
-//
-void cullOverReferences() {
-
-  /* A note about the structure of this pass:
 
-     While it's interprocedural, it shouldn't present a big problem for
-     separate compilation. The reason is that functions compiled into
-     a library should know whether or not they are setting their arguments.
-     Combining that fact with the plan to make the compiler produce a
-     "header file" in addition to "implementation files" as part of a library,
-     there is no problem.
-
-     (However, any interface describing a callable function / call-back
-      function will need to have some other default -- possibly blank
-      argument intent leads to compilation error for ref-if-modified types
-      like arrays).
+// TODO: Move this into `resolution/loopDetails.cpp` and adjust the function
+// `gatherLoopDetails` to use it.
+struct LoopInfo {
+  bool isForall;
+  IteratorDetails leaderDetails;
+  ForLoop* followerForLoop;
+  std::vector<IteratorDetails> detailsVector;
+};
 
-     Second, this pass works in a uses-to-symbols manner and is an
-     interprocedural analysis. However, that does not mean that it can't
-     fit into a Pass Manager concept. In particular, it could be
-     what LLVM calls a CallGraphSCCPass.
-  */
+struct CullRefCtx {
+public:
+
+  void propagateConstnessToMethodReceivers(void);
+  void propagateConstnessToFieldAccesses(void);
+  void collectTuplesAndRefMaybeConstArgs(void);
+  void collectOrLowerContextCallExprs(void);
+  void visitCollectedSymbols(void);
+
+  bool isCallApplicableLoop(SymExpr*, CallExpr*, GraphNode, LoopInfo&);
+  bool checkLoopForDependencies(GraphNode, LoopInfo&, bool&);
+  bool checkBuildTupleCall(SymExpr*, CallExpr*, GraphNode, bool&);
+  bool checkContextCallExpr(CallExpr*, GraphNode, bool&);
+  bool checkTupleCastCall(SymExpr*, CallExpr*, GraphNode, bool&);
+  bool checkAccessorLikeCall(SymExpr*, CallExpr*, GraphNode, bool&);
+  bool checkRefMaybeConstFormal(SymExpr*, CallExpr*, GraphNode, bool&);
+  bool checkGetRefTupleField(CallExpr*, GraphNode, bool&);
+  bool checkSetRefTupleField(CallExpr*, GraphNode, bool&);
+  bool checkCompilerRefTemporaries(CallExpr*, GraphNode, bool&);
+  bool checkOuterVariableInForall(SymExpr*, GraphNode, bool&);
+  void visitCollectedSymbol(GraphNode);
+
+  void propagateNonConstnessThroughRevisitGraph(void);
+  void markRemainingSymsWithUnknownConstnessConst(void);
+  void lowerRemainingContextCallExprs(void);
+  void performLateConstCheck(void);
+
+private:
```

Comment:
I don't personally see the point of making anything private for a class that only can be used within one file, but that's just a style thing

Reply:
Yeah, I couldn't make up my mind on whether to make it a class or a struct. I'll keep that in mind for next time.